pinecone/data/index.py

```diff
@@ -502,6 +502,40 @@ def describe_index_stats(
             ),
             **{k: v for k, v in kwargs.items() if k in _OPENAPI_ENDPOINT_PARAMS},
         )
+    
+    @validate_and_convert_errors
+    def list_paginated(
+        self,
+        prefix: Optional[str] = None,
+        limit: Optional[int] = None,
+        pagination_token: Optional[str] = None,
+        namespace: Optional[str] = None,
+        **kwargs
+    ) ->  ListResponse:
+        args_dict = self._parse_non_empty_args(
+            [
+                ("prefix", prefix),
+                ("limit", limit),
+                ("namespace", namespace),
+                ("pagination_token", pagination_token),
+            ]
+        )
+        return self._vector_api.list(**args_dict, **kwargs)
+
+    @validate_and_convert_errors
+    def list(self, **kwargs):
+        limit = kwargs.get("limit", 100)
+        done = False
+        while not done:
+            results = self.list_paginated(**kwargs)
+            if len(results.vectors) > 0:
+                yield [v.id for v in results.vectors]
+            
+            full_page = len(results.vectors) == limit
+            if results.pagination and full_page:
```

Comment:
@jhamon it's valid for a page to contain less than `limit` results without being the last page. While I do not see that happening with List, I prefer to leave this freedom to engineering.
Therefore, the only indication of whether we've reached the end is whether the result has any pagination token.

Reply:
What you're describing is the current behavior, but I don't think it's valid. I reported it as a bug and your team filed an Asana issue for it. I added this "full page" check so that the behavior would be deterministic in testing and not fire off additional calls to fetch an empty array.

Reply:
The bug was specifically about adding an extra empty page at the end of the pagination loop, but the generic case of sending <limit results is valid - see the [API Pagination PRD](https://docs.google.com/document/d/19lJBFhJ2VqBSGwHzenUk40Oe64mPf3d0QttY2PP-tiQ/edit#heading=h.hgq4984r4biv):
> Users should not assume that a page containing fewer results than the limit value is necessarily the last one - and instead should always check the pagination.next field.

Reply:
I agree with Ben here. In the generic case we should have the ability to shards result and than it is common practice to return less than the full page and still have continuation token.